scripts/sbatch.py

```diff
@@ -0,0 +1,476 @@
+    """
+    keyword="Shared="
+    #NOTE: this keyword may change if NeSI updates the slurm version
+    #"OverSubscribe" instead of "Shared"
+    try:
+        value = get_value_from_conf(slurm_conf, partition_name, keyword)
+        if value.upper() == "EXCLUSIVE":
+            return False
+        else:
+            return True        
+    except ValueError:
+        print("no shared definition found, assuming TRES are shared")
+        return True
+    
+
+def calculate_requested_chours(
+        cpu_billing_weights,
+        mem_billing_weights,
+        total_seconds_requested,
+        mem_per_cpu,
+        ntasks,
+        ntasks_per_core,
+        cpus_per_task,
+        shared,
+        cpu_per_node,
+        mem_per_node,
+        ntasks_per_node,
+        nodes=None,
+):
+    """Calculate requested core hours for a job """
+    # https://slurm.schedmd.com/tres.html
+    # see above link for calculating formula
+    if (cpus_per_task is not SlurmHeader.cpus_per_task.default):
+        total_cpus = ntasks * cpus_per_task
+    else:
+        total_cpus = ntasks / ntasks_per_core
+    if shared == False:
+        nodes = ceil(total_cpus/cpu_per_node)
+        if ntasks_per_node is not SlurmHeader.ntasks_per_node.default:
+            nodes = nodes * (ntasks/ntasks_per_node)
+        total_cpus = nodes * cpu_per_node
+        total_mem  = nodes * mem_per_node
+    else:
+        total_mem = total_cpus * mem_per_cpu
+    #TODO:GPU may need to be included
+    requested_hours = (total_seconds_requested * (total_cpus * cpu_billing_weights)) + (
+        total_mem * mem_billing_weights
+    ) 
+    return requested_hours
+
+
+def get_available_chours(json_file, account, username):
+    """
+       Read the core hours json file that's populated from the dashboard.db and
+       return the available core hours for a specified user in specified account
+    """
+    with open(json_file, "r") as f:
+        json_array = json.load(f)
+    for d in json_array:
+        if d.get(account) is not None:
+            for user_dict in d[account]:
+                if user_dict.get(username) is not None:
+                    return (
+                        user_dict[username]["allocation"] - user_dict[username]["used"]
+                    )
+            #no data for a user is found while account exist
+            sys.exit(
+                "No core hours usage info for {} {} from json file {}".format(
+                    account, username, json_file))
+    #all project names looped raise error
+    raise IndexError("No account info for {} from json file {}".format(account, json_file))
+
+
+def process_slurm_header(sl_file):
+    """Read a slurm file and returns the wanted header values as a dictionary"""
+    with open(sl_file) as f:
+        lines = f.readlines()
+    #example line: SBATCH --header=value
+    header_dict = {h.value: h.default for h in SlurmHeader}
+    for line in lines:
+        if line.upper().startswith("#SBATCH"):
```

Comment:
short circuit rather than checking the entire file?

Reply:
how do we determine when to stop? any non-blank-line or not-start-with-hash?

Reply:
Yep. That's what I'd do. I'd almost say that an SBATCH directive after the header is bad news anyway.